src/Compilers/Core/Portable/Diagnostic/CommonMessageProvider.cs

```diff
@@ -13,6 +13,11 @@ namespace Microsoft.CodeAnalysis
     /// </summary>
     internal abstract class CommonMessageProvider
     {
+        /// <summary>
+        /// Caches the return values for <see cref="GetIdForErrorCode(int)"/>.
+        /// </summary>
+        private static readonly ConcurrentDictionary<(string prefix, int code), string> s_errorIdCache = new ConcurrentDictionary<(string prefix, int code), string>();
```

Comment:
Do you have any idea on the number of unique codes this path was seeing?

Reply:
It would be at most one for each error code produced by the C# and VB compilers (the built-in codes that are reported with the CS or BC prefix).

Reply:
Would interlocked operations on an ImmutableOperations be a better option than concurrent dictionary?

Reply:
I would not expect `ImmutableDictionary<TKey, TValue>` to provide a performance advantage here.